src/Language/Haskell/TH/Datatype.hs

```diff
@@ -316,37 +317,120 @@ normalizeDec' context name params cons variant =
        }
 
 -- | Normalize a 'Con' into a 'ConstructorInfo'. This requires knowledge of
--- the type and parameters of the constructor as extracted from the outer
+-- the type and parameters of the constructor, as well as whether the constructor
+-- is for a data family instance, as extracted from the outer
 -- 'Dec'.
 normalizeCon ::
-  Name   {- ^ Type constructor -} ->
-  [Type] {- ^ Type parameters  -} ->
-  Con    {- ^ Constructor      -} ->
+  Name            {- ^ Type constructor  -} ->
+  [Type]          {- ^ Type parameters   -} ->
+  DatatypeVariant {- ^ Extra information -} ->
```

Comment:
It's a bit unfortunate I had to change the public API here to account for this eta-reduction bug, but I don't see a cleaner way to tell if we're in a data family instance or not within `normalizeCon`. We could re-do all of the work in `normalizeDec`, but that feels rather gross...